Hud icons are now based on ID card styles (+crew monitor, crew manifest are based on hud icons than job names + chameleon agent card QoL)

EvilDragonfiend opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request
tracking Hud icon based on ID card style

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game
a basic start to implementing alternative job title system.
Also, you might want to give a crew a special job, but seeing them as a grey unknown hud icon is kinda annoying.


## Testing Photographs and Procedure
<!-- Include any screenshots/videos/debugging steps of the modified code functioning successfully, ideally including edge cases. -->
<details>
<summary>Screenshots&Videos</summary>

https://youtu.be/qMRL6g63jWA

(ignore some weird manipulations. They're fixed.)

![image](https://user-images.githubusercontent.com/87972842/175192667-cd1b884d-935d-4979-bd32-02c6bbaa37d3.png)

Job title is Captain, but HUD is security officer because it's Sec card style.


![image](https://user-images.githubusercontent.com/87972842/175203433-cf599a75-65af-46d3-a30a-0efcfc03bd62.png)

crew monitor tracks based on hud icon now


![image](https://user-images.githubusercontent.com/87972842/175813480-8233b771-cdd2-4bdf-875a-34a321d0d4e8.png)
![image](https://user-images.githubusercontent.com/87972842/175813891-bbcc5623-77e0-462c-98e6-751f6d93dc9c.png)

crew manifest is based on hud icon department type too (regardless of job name)



-------------

![image](https://user-images.githubusercontent.com/87972842/176335255-f45e260f-5d66-4fb1-89ca-748d5d46dacd.png)

chameleon QOL
Once you change a card into a job card, you'll get a pop up, and if you click yes, you get the default job title of the chosen card.
if there's no default name or you chose no, you'll still get the previous job title.

--------
https://www.youtube.com/watch?v=b6RU4wTuUv0&feature=youtu.be

broken chameleon agent card


</details>

## Changelog
:cl: EvilDragonfiend, PestoVerde322
add: Hud is now tracked based on your card appearance. HoP is now supposed to repaint crews' card in order to give them correct hud icons, or securities(+ghosts) will blame you.
add: every card now has default hud icons. (i.e. Captain's Spare card has a blank command HUD, budget cards have blank department color HUD)
add: added more PDA style options to PDA painter. (deputy, prisoner, heads, unassigned, etc)
refactor: PDA painter is now easy to recognise colourising style easier.
tweak: Some job HUDs are now only available through emaged ID card repainter.
tweak: crew monitor now tracks crew job color based on hud state.
tweak: Assistant id card crew objective is now greentextable when your card is painted even if the job name is still assistant.
add: hud image based defines
refactor: crew manifest is based on hud state.
refactor: PDA painter machine now updates the crew manifest upon ID repainting
refactor: changes to chameleon code and agent card, to make Hud icon update working. Chameleon agent card has earned QoL. They'll get a default job name of a card type you chose once changed.
add: agent card will be randomised upon EMP. (random hud icon, random job name, random icon)
add: broken chameleon agent card (with all wiped access, and unable to copy accesses from another card)
add: broken chameleon kit will have broken chameleon agent card.
add: new hud icons, new card images (blank symbol department color hud icons, new hudking icon)
del: 9x9 pixel size 'hudking' hud icon
add: New card types for the purpose of chameleon agent card. (i.e. "their majesty's gold identification card")
tweak: Some card types are now blacklisted from chameleon agent card - because they don't make sense. (i.e changeling fake ones)
tweak: your card paycheck department is now re-assigned based on your hud icon state. if your hud icon is red with job name 'not real security', you are paid by the security budget. the process happens when you repaint a card or give a new job name to a card.
tweak: not all gimmicks are in civilian department anymore. Psychiatrist is medical, and Stage Magician is service like their mime/clown fellows. (VIP and barber are still service)
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

Commit:
hudicons

Commit:
crew monitor

Commit:
badif

EvilDragonfiend commented on code/game/machinery/computer/crew.dm:
```diff
@@ -190,10 +198,10 @@ GLOBAL_DATUM_INIT(crewmonitor, /datum/crewmonitor, new)
-			if(jobs[I.assignment] != null)
-				entry["ijob"] = jobs[I.assignment]
+			entry["name"] = I.registered_name ? I.registered_name : "Unknown"
+			entry["assignment"] = I.assignment ? I.assignment : "Unknown"
```
card with no name/assignment makes bad null bug here.

EvilDragonfiend commented on code/game/objects/items/cards_ids.dm:
```diff
@@ -345,45 +357,19 @@ update_label("John Doe", "Clowny")
+	hud_state = "syndicate"
 	var/anyone = FALSE //Can anyone forge the ID or just syndicate?
 	var/forged = FALSE //have we set a custom name and job assignment, or will we use what we're given when we chameleon change?
-	var/static/list/available_icon_states = list(
```
actually not used variable

Commit:
aa

Commit:
a

Commit:
paint counts

Crossedfall commented:
Good change

Penwin0 commented on code/game/machinery/PDApainter.dm:
```diff
@@ -43,38 +43,25 @@
+
+	var/valid_jobs = list(
+		// Command
+		"----Command----","Command (Custom)", "Captain", "Acting Captain",
```
Won't this show "----Command----" on the list of options in the PDA painter? What happens if I try to pick it?

EvilDragonfiend replied:
Does nothing. It's visual boundaries for the selection window.

Penwin0 commented on code/game/machinery/PDApainter.dm:
```diff
@@ -267,3 +261,155 @@
 	update_icon()
+
+
+/proc/get_cardstyle_by_jobname(jobname)
```
I feel like both `get_cardstyle_by_jobname` and `get_hud_by_jobname` could both be assimilated into one, there's a lot of repetition, or just handled in `var/valid_jobs` like it was previously

EvilDragonfiend replied:
likely no.
```
(card style)
			"Botanist" = "serv",
			"Bartender" = "serv",
			"Cook" = "serv",
```
```
(hud style)
			"Bartender" = "bartender",
			"Cook" = "cook",
			"Botanist" = "botanist",
```
each of them has different values and they can't be put together. syncing their names isn't a good idea, and putting them into a single proc will make this more uglier as well, 
I don't like this too, but I don't have any particular idea to make it better pretty.

EvilDragonfiend replied:
I forgot to put another reason too.
`get_hud_by_jobname`
`get_cardstyle_by_jobname`
proc names are obvious to realise what they do.

Commit:
unknownhud

Commit:
minor bugfix

Commit:
typo

ItsTisme commented:
if your job is manualy changed it should prop update the card/icon(assistant being promoted to medical or jannitor as an example

EvilDragonfiend commented:
> if your job is manualy changed it should prop update the card/icon(assistant being promoted to medical or jannitor as an example

I thought about it, but I want HoP to be sure about the importance of the painting machine, and make the field promotion less usual.

Commit:
better chameleon

EvilDragonfiend commented:
currently touching crew mamifest too

Commit:
big changes


ff


cc


dd


gg


rwer

EvilDragonfiend commented:
new change. now crew manifest is based on hud

<details>
<summary>Screenshots&Videos</summary>

![image](https://user-images.githubusercontent.com/87972842/175813480-8233b771-cdd2-4bdf-875a-34a321d0d4e8.png)
![image](https://user-images.githubusercontent.com/87972842/175813891-bbcc5623-77e0-462c-98e6-751f6d93dc9c.png)

crew manifest is based on hud icon department type too (regardless of job name)

</details>


Now this PR might be ready to take a full review now.

Commit:
minorfixes

Commit:
mf2

Commit:
self important

Commit:
chameleon eddition

EvilDragonfiend commented:
there will be no additional change from now. might be ready to take a review.

Commit:
tidy

PowerfulBacon reviewed (commented):
What happens if I wear a budget card or other card that isn't an ID?

Commit:
king

Commit:
missing hud

Commit:
PDA

add: added more PDA style options to PDA painter. (deputy, prisoner, heads, unassigned, etc)

EvilDragonfiend committed:
Merge branch 'master' into altaccess

Commit:
relocate some codes into `_define/job.dm`

Commit:
Update your Bank based on hud_state

now card paycheck department is based on hud_state, not the given job.

Commit:
bug fix

Commit:
better comment

Commit:
minor fix

PowerfulBacon reviewed (approved):
Yolo, seems alright
